app/services/control_rate_service.rb

```diff
@@ -10,60 +10,60 @@ def initialize(region, periods:)
     @region = region
     @facilities = region.facilities
     # Normalize between a single period and range of periods
-    @periods = if !periods.is_a?(Range)
+    @report_range = if !periods.is_a?(Range)
       # If calling code is asking for a single period,
       # we set the range to be the current period to the start of the next period.
       Range.new(periods, periods.succ)
     else
       periods
     end
-    @quarterly_report = @periods.begin.quarter?
-    @results = Reports::Result.new(@periods)
+    @quarterly_report = @report_range.begin.quarter?
+    @results = Reports::Result.new(region: @region, period_type: @report_range.begin.type)
     logger.info class: self.class, msg: "created", region: region.id, region_name: region.name,
-                periods: periods.inspect, facilities: facilities.map(&:id)
+                report_range: report_range.inspect, facilities: facilities.map(&:id)
   end
 
   delegate :logger, to: Rails
   attr_reader :facilities
-  attr_reader :periods
   attr_reader :region
+  attr_reader :report_range
   attr_reader :results
 
+  # We cache all the data for a region to improve performance and cache hits, but then return
+  # just the data the client requested
   def call
-    Rails.cache.fetch(cache_key, version: cache_version, expires_in: 7.days, force: force_cache?) do
-      results.registrations = registration_counts
-      results.cumulative_registrations = sum_cumulative_registrations
-      results.count_adjusted_registrations
-
-      periods.each do |(period, count)|
-        results.controlled_patients[period] = controlled_patients(period).count
-        results.uncontrolled_patients[period] = uncontrolled_patients(period).count
-      end
-
-      results.calculate_percentages(:controlled_patients)
-      results.calculate_percentages(:uncontrolled_patients)
-      results
-    end
+    all_cached_data.report_data_for(report_range)
```

Comment:
This is the major change here: we grab all the cached data and then ask the result to filter it to just the report range.